### PR TITLE
Register/implement golden functions for the comparison mode

### PR DESCRIPTION
**Pull request by ctr-martemovTT**
### PR Category

Part of the implementation for https://github.com/tenstorrent/tt-metal/issues/28214 -- focus on providing the implementation of the golden functions for the un-annotated ttnn operations. 

Follow-up for https://github.com/tenstorrent/tt-metal/pull/55623

### Summary

- add golden functions for the operations missing the implementation
- add support for per-operation output name overrides, used in the moreh operations

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:martemov/comparison-mode-fixes-4)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:martemov/comparison-mode-fixes-4)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:martemov/comparison-mode-fixes-4)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:martemov/comparison-mode-fixes-4)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:martemov/comparison-mode-fixes-4)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:martemov/comparison-mode-fixes-4)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:martemov/comparison-mode-fixes-4)

### CODE REVIEW

**Comment by pavlejosipovic** [comment]
For the conv/pool/SDPA-owned area, can we please avoid introducing duplicate reference implementations and first reuse/consolidate what is already present?

- `ttnn.conv2d` already has an attached golden in `ttnn/ttnn/operations/conv2d.py`.
- The public pooling operations already have attached goldens in `ttnn/ttnn/operations/pool.py` (`max_pool2d`, `avg_pool2d`, global/adaptive pools, `rotate`, `upsample`, and `grid_sample`).
- SDPA does not currently have these references attached to the operations, but reference implementations already exist in places such as `tests/ttnn/unit_tests/operations/sdpa/sdpa_test_utils.py` and `tests/sweep_framework/sweeps/model_traced/paged_scaled_dot_product_attention_decode_model_traced.py`.

Please check whether those existing references can be moved to a shared production module and attached, instead of maintaining another independent implementation in `operations/transformer.py`. Otherwise the test oracle and comparison-mode golden can drift independently. `conv1d` and `conv_transpose2d` do appear to be genuinely missing registered goldens.

File: ttnn/ttnn/operations/transformer.py [hunk 1]
```diff
@@ -133,4 +133,643 @@ def apply_rotary_pos_emb(x, cos_cached, sin_cached, token_idx=0):
...
+        input_tensor_v,
+        attn_mask=attn_mask,
+        is_causal=is_causal,
+        scale=scale,
+        sliding_window_size=sliding_window_size,
+        attention_sink=attention_sink,
+    )
+
+
+ttnn.attach_golden_function(ttnn.transformer.scaled_dot_product_attention, golden_function=_golden_function_sdpa)
```
Comment (pavlejosipovic) [inline ttnn/ttnn/operations/transformer.py]: Can we avoid duplicating reference implementations here and consolidate what already exists? In the conv/pool/SDPA area, `ttnn.conv2d` already has an attached golden, and the public pooling operations already have attached goldens in `ttnn/ttnn/operations/pool.py`. SDPA is not currently attached, but reference implementations already exist in `tests/ttnn/unit_tests/operations/sdpa/sdpa_test_utils.py` and `tests/sweep_framework/sweeps/model_traced/paged_scaled_dot_product_attention_decode_model_traced.py`. Could those references be moved into a shared production module and attached here, rather than maintaining a second independent implementation that can drift from the test oracle? `conv1d` and `conv_transpose2d` do appear to be genuinely missing registered goldens.

Comment (ctr-martemovTT) [reply ttnn/ttnn/operations/transformer.py]: @pavlejosipovic requested changes will notably increase the scope/size of this PR, so I decided to instead revert the changes to the conv2d, SDPA etc. and will create a follow-up PR after, targeting specifically this part, taking into account your feedback.

Comment (pavlejosipovic) [reply ttnn/ttnn/operations/transformer.py]: pr title is that you are adding missing golden impls, they are not missing, maybe you can help me understand what is the goal here?

Comment (ctr-martemovTT) [reply ttnn/ttnn/operations/transformer.py]: @pavlejosipovic the objective for this PR is to check which operations don't have the golden functions and implement them. I primarily used the `ttnn. dump_operations` function that reports if the operation has a golden function attached or not. 

`dump_operations()` reports Has `golden function = False` whenever no callable is attached to the registered operation, even when tests contain an oracle. So the confusion was mostly in terminology on my end, I meant "missing" as "not attached/visible to the comparison mode", not as in "completely missing from the repository". 

I reverted some of the changes because I wanted to keep the scope of this PR closer to "for golden functions actually have no reference implementation, provide the implementation and attach it to the ttnn operations". 

I fully agree that the reference implementations from the test oracle should be reused, I'm only going for the separate follow-up PR to make the review easier: one low-complexity PR for removing goldens  that are truly missing, and another one for re-organizing the existing reference implementations and attaching them to the operations.

File: tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_getitem.py [hunk 2]
```diff
@@ -25,6 +25,18 @@ def to_output_5d_shape(shape, index_dims, index_size):
...
+def test_moreh_getitem_golden_uses_paired_negative_indices():
+    input_tensor = torch.arange(24).reshape(2, 3, 4)
+    first_index = torch.tensor([0, -1])
+    second_index = torch.tensor([1, 2])
+    golden_function = ttnn.get_golden_function(ttnn.moreh_getitem)
+
+    actual = golden_function(input_tensor, [first_index, second_index], [1, 2])
+    expected = input_tensor[:, first_index, second_index]
+
+    torch.testing.assert_close(actual, expected)
```
Comment (nmauriceTT) [inline tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_getitem.py]: We should assert against equality here.

File: ttnn/ttnn/operations/unary.py [hunk 3]
```diff
@@ -3,148 +3,160 @@
...
+    def torch_hardmish(x):
+        x_f32 = x.to(torch.float32)
+        result_f32 = x_f32 * torch.clamp(x_f32 * 0.5 + 1.0, min=0.0, max=1.0)
+
+        if x.dtype == torch.bfloat16:
+            # Simulate SFPSTORE truncating
+            result_int32 = result_f32.view(torch.int32)
+            shifted_int32 = torch.bitwise_right_shift(result_int32, 16)
+            truncated_int16 = shifted_int32.to(torch.int16)
+            final_result = truncated_int16.view(torch.bfloat16)
```
Comment (nmauriceTT) [inline ttnn/ttnn/operations/unary.py:34]: These are supposed to be golden functions right ? Why is it trying to reproduce `SFPSTORE` behavior ? 

If ttnn's implementation of hardmish is buggy then let's fix it, not the golden.

Comment (ctr-martemovTT) [reply ttnn/ttnn/operations/unary.py:34]: @nmauriceTT This implementation predates the PR -- the code exists in main https://github.com/tenstorrent/tt-metal/blob/3815ba4e706ee04b4b2163fe088b46854bdb244c/ttnn/ttnn/operations/unary.py#L24-L37

It should be possible to fix this mismatch, but not in the scope of this PR -- it is primarily for adding goldens that are missing. 

I can take this is a todo item for the follow-up PR, to avoid expanding the scope of changes here. 

<details>
Initial investigation with cursor suggest the fix is not complex, but requires changes to the kernels:

Underlying reason: HardMish computes in FP32, but when BF16 DEST is used, `SFPSTORE` truncates the lower 16 bits instead of IEEE round-to-nearest-even. PyTorch rounds, so the golden manually truncates to match hardware. FP32 outputs are unaffected; Wormhole and Blackhole behave similarly.

The actual fix is small:
- Pass `DST_ACCUM_MODE` through `hardmish_tile`.
- Before storing BF16, apply `sfpi::convert<sfpi::vFloat16b>(result, sfpi::RoundMode::Nearest)`.
- Update both Wormhole and Blackhole kernels.
- Simplify the Python golden and adjust its accuracy-test documentation.

This pattern already exists in `mish`, `cbrt`, and other SFPU kernels.
</details>

Comment (nmauriceTT) [reply ttnn/ttnn/operations/unary.py:34]: Yes. Sounds good as a follow-up item.

File: tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_getitem.py (hunk 2 above)
Comment (ctr-martemovTT) [reply tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_getitem.py]: **resolved**: updated to `assert torch.equal(actual, expected)`

File: ttnn/ttnn/operations/transformer.py (hunk 1 above)
Comment (pavlejosipovic) [reply ttnn/ttnn/operations/transformer.py]: these are not from test oracle these are what is used for tech_reports/ttnn/comparison-mode.md

**Review by sjameelTT — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]
Please create a follow up issue for fixing up moreh and assign it to @iwroszTT for ign to look at. It seems like some of the Moreh ops like adam, sum, and mean will need to be updated.

**Review by nmauriceTT — APPROVED** [review]

File: ttnn/ttnn/operations/transformer.py (hunk 1 above)
Comment (ayerofieiev-tt) [reply ttnn/ttnn/operations/transformer.py]: I think its fine to have separate changes. 
Sure agree with @pavlejosipovic to avoid duplication.

**Review by ayerofieiev-tt — APPROVED** [review]

**Comment by p32929** [comment]
@ctr-martemovTT — this PR rewrites exactly the function I just filed #57115 against, so flagging it here rather than letting the two land independently.

The new `_ReductionGoldenSpec` factory fixes `correction`, the `torch.max/min(dim=...)` namedtuple, and `argmax` under `dim=None` — but `scalar=` is still absorbed by `**_`:

```python
def golden_function(input_tensor, dim=None, keepdim=False, correction=None, **_):
```

`scalar` is a live, non-deprecated parameter on all six of these ops (`generic_reductions_nanobind.hpp:179`, `std_var_reductions_nanobind.cpp:82`/`:97`), documented as *"A scaling factor to be applied to the input tensor"*, and `generic_reductions.cpp` threads it into the reduce — including the sign-flip on lines 736-745. So the golden currently returns the unscaled reduction whenever a caller passes one.

I ran both variants (today's `main` and this PR's rewrite) side by side in pure Python; the divergence is identical in both — up to 300% relative error on `var(scalar=0.5)`, and `max`/`min` with a negative scalar return a *different element* than the device. Full table and a runnable self-checking repro are in #57115.

It looks like one more spec field plus two lines here:

```python
def golden_function(input_tensor, dim=None, keepdim=False, correction=None, scalar=1.0, **_):
    if scalar != 1.0:
        input_tensor = input_tensor * scalar
```

which reproduces all five existing `scalar=` test contracts (`test_max.py:164`, `test_reduction.py:132`/`:180`, `test_reduction_mean.py:99`, `test_sum_int.py:106`), negative-scalar flip included, since `amax(-2*x) == -2*min(x)`.

Happy to fold it into this PR or send it separately — whichever you prefer.

**Comment by ctr-martemovTT** [comment]
@p32929 in this case the more optimal solution seems to handle the two PRs independently, there is no difference between (1) merging the fix into this PR, (2) merging this PR into main and then rebasing a fix PR, (3) merging a fix PR and rebasing this branch. Since the fix addresses pre-existing golden function state, I would prefer to keeps the PR separate. 

Additional question: I noticed you have created several issues for golden functions https://github.com/tenstorrent/tt-metal/issues?q=is%3Aissue%20state%3Aopen%20author%3Ap32929 -- is it a part of ongoing task for reviewing the existing golden functions?

File: ttnn/ttnn/operations/transformer.py (hunk 1 above)
Comment (ctr-martemovTT) [reply ttnn/ttnn/operations/transformer.py]: I'm considering the question resolved in the scope of this PR -- handling of these functions will be done in a separate set of changes in the follow-up PR, and will reuse the existing implementations.
